nifi-toolkit/nifi-toolkit-zookeeper-migrator/src/main/java/org/apache/nifi/toolkit/zkmigrator/ZooKeeperMigrator.java

```diff
@@ -284,8 +286,40 @@ private Stat transmitNode(ZooKeeper zooKeeper, DataStatAclNode node) {
     }
 
     private ZooKeeper getZooKeeper(ZooKeeperEndpointConfig zooKeeperEndpointConfig, AuthMode authMode, byte[] authData) throws IOException {
+        CountDownLatch connectionLatch = new CountDownLatch(1);
         ZooKeeper zooKeeper = new ZooKeeper(zooKeeperEndpointConfig.getConnectString(), 3000, watchedEvent -> {
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("ZooKeeper server state changed to {} in {}", watchedEvent.getState(), zooKeeperEndpointConfig);
+            }
+            switch (watchedEvent.getType()) {
+                case None:
+                    switch (watchedEvent.getState()) {
+                        case SyncConnected:
+                            connectionLatch.countDown();
+                            break;
+                        case Expired:
+                        case AuthFailed:
+                        case ConnectedReadOnly:
+                        case SaslAuthenticated:
+                        case Disconnected:
+                            break;
+                    }
+            }
         });
+
+        final boolean connected;
+        try {
+            connected = connectionLatch.await(5, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException(String.format("interrupted while waiting for ZooKeeper connection to %s", zooKeeperEndpointConfig), e);
```

Comment:
However, there are other areas in the code that throw RuntimeExceptions.  Could address those in another PR if necessary.